Subscription management: add "Add sites" button & modal skeleton

TimBroddin opened the pull request:
Closes #79226

## Proposed Changes

This PR:

- Adds a `<AddSitesButton />`,  `<AddSitesModal />` & `<AddSitesForm />`
- Implements the button inside Reader
- Adds some groundwork for future PR's (such as passing needed props to AddSitesForm)

The button will only show when the feature flag `subscription-management-add-sites-modal` is set. This feature flag is enabled for development.

## Testing Instructions

- Apply this PR
- Visit `http://calypso.localhost:3000/read/subscriptions`
- You should see the `Add sites` button
- Clicking it should show the skeleton modal

![CleanShot 2023-07-12 at 11 37 32@2x](https://github.com/Automattic/wp-calypso/assets/528287/bbff7256-17f2-4366-93e8-d9c4e26a5d2f)

- Disable the feature flag in `config/development.json` (line 193)
- Restart Calypso
- The button should not be shown.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

TimBroddin committed:
Add subscription-management-add-sites-modal feature flag

TimBroddin committed:
Add sites button & skeleton modal

TimBroddin committed:
Only show the button & modal if 'subscription-management-add-sites-modal' feature flag is enabled

In the 'AddSitesButton' component, the feature flag 'subscription-management-add-sites-modal' is checked before rendering. If the feature is disabled, component will not be shown. This addition provides the ability to control the display of the 'AddSitesButton' and its associated modal, across different deployment environments.

ivan-ottinger commented:
I have started looking at this PR and noticed that the Unit tests failed because of the following error:

```
client/reader/site-subscriptions-manager/site-subscriptions-manager.tsx(5,2): error TS2724: '"@wordpress/components"' has no exported member named '__experimentalSpacer'. Did you mean '__experimentalHStack'?
```

TimBroddin commented:
> I have started looking at this PR and noticed that the Unit tests failed because of the following error:
> 
> ```
> client/reader/site-subscriptions-manager/site-subscriptions-manager.tsx(5,2): error TS2724: '"@wordpress/components"' has no exported member named '__experimentalSpacer'. Did you mean '__experimentalHStack'?
> ```

That's strange, I also got that error but the component renders 🤷‍♂️ Let me look into it.

TimBroddin committed:
Add __experimentalSpacer in wp-components-types

ivan-ottinger commented on client/landing/subscriptions/components/add-sites-button/add-sites-button.tsx:
```diff
@@ -0,0 +1,34 @@
+				className="subscriptions-add-sites-button"
+				onClick={ () => setIsAddSitesModalVisible( true ) }
+			>
+				{ translate( 'Add sites' ) }
```
Just a note (so we won't forget): If we agree to display form for one site only, this will likely need to change to `Add site` (and similarly in the modal's title).

TimBroddin replied:
I think it can stay like it is, because when uploading an OPML file, you're still adding multiple sites.

ivan-ottinger replied:
Good point. 👍🏼

ivan-ottinger commented on client/reader/site-subscriptions-manager/site-subscriptions-manager.tsx:
```diff
@@ -1,6 +1,8 @@
 import {
 	__experimentalHStack as HStack,
 	__experimentalVStack as VStack,
+	// eslint-disable-next-line wpcalypso/no-unsafe-wp-apis
+	__experimentalSpacer as Spacer,
```
I wonder why the two imported values above (`HStack` and `VStack`) don't need the eslint comment as well. 🤔

ivan-ottinger replied:
Oh, does this @mashikag's https://github.com/Automattic/wp-calypso/pull/79289#issuecomment-1632551804 explain that? 🙂

TimBroddin replied:
Yes, I'll push an update :)

mashikag commented:
> That's strange, I also got that error but the component renders 🤷‍♂️ Let me look into it.

It is just missing type definition, look at how we define temp type defs for other experimental componnets. 🙂

TimBroddin commented:
> > That's strange, I also got that error but the component renders 🤷‍♂️ Let me look into it.
> 
> It is just missing type definition, look at how we define temp type defs for other experimental componnets. 🙂

I added it, still running the tests 😵‍💫

mashikag commented:
Oh, you should also just add an exception in the [.eslintrc.js](https://github.com/Automattic/wp-calypso/blob/55414496b4734adf16b5255b2de3f4d3e7ed54ed/.eslintrc.js#L492) .

ivan-ottinger reviewed (approved):
Looks great and works as expected. 🚀 Very nice and clean code. 

- I did not notice any issues, errors, warnings, nor regressions
- all strings are run through the `translate` function
- the button displays only when the feature flag is enabled 👍🏼 

Just shared a couple comments / observations below. 🙂

TimBroddin committed:
In .eslintrc.js, __experimentalSpacer was added to the list of allowed components in order to avoid unnecessary eslint warnings.

TimBroddin commented:
> Oh, and you should also just add an exception in the [.eslintrc.js](https://github.com/Automattic/wp-calypso/blob/55414496b4734adf16b5255b2de3f4d3e7ed54ed/.eslintrc.js#L492) . Otherwise a warning will be displayed. 🙂

Thanks for the pointer! Makes sense to put it there instead of inline-disabling the rule.

ivan-ottinger commented on client/wp-components-types.d.ts:
```diff
@@ -8,6 +8,8 @@ declare module '@wordpress/components' {
 	export const __experimentalDivider: React.ComponentType< Props >;
 	export const __experimentalHStack: React.ComponentType< Props >;
 	export const __experimentalVStack: React.ComponentType< Props >;
+
```
Looks like extra line here.

TimBroddin replied:
Oops! Fixed it. Thanks for catching it!

TimBroddin committed:
Remove unnecessary whitespace in wp-components-types

This commit removes a superfluous line in the wp-components-types.d.ts file, enhancing readability and clean code practices. This doesn't affect functionality but aids maintainability.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8544639

Thank you @TimBroddin for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8544639).
